nemo/utils/callbacks/ipl_epoch_stop.py

```diff
@@ -0,0 +1,39 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from lightning.pytorch import Trainer
+from lightning.pytorch.callbacks import Callback
+from lightning.pytorch.core import LightningModule
+
+
+class IPLEpochStopper(Callback):
+    r"""
+    Gracefully terminates training at the *end* of an epoch.
+    This is done to generate pseudo-labels dynamically.
+    enable_stop : bool, default=False
+        If ``True`` the callback will request a stop in
+        :py:meth:`on_train_epoch_end`.  If ``False`` it is inert.
+    """
+
+    def __init__(self, enable_stop: bool = False) -> None:
+        super().__init__()
+        self.enable_stop = bool(enable_stop)
+
+    def on_train_epoch_end(self, trainer: Trainer, pl_module: LightningModule) -> None:
+        """
+        Sets `should_stop` stop flag to terminate the training.
+        """
+        super().__init__()
+        if self.enable_stop:
+            trainer.should_stop = True
```

Comment:
Thanks, move this to https://github.com/NVIDIA/NeMo/tree/main/nemo/collections/common/callbacks (lot of similar folders :D , this seemed to be best one for this usecase)

Then add this through exp_manager. See: https://github.com/NVIDIA/NeMo/blob/6e7b414dfcd4ec955278755c24062e66f5ac848e/nemo/utils/exp_manager.py#L708 and you can add similarly for this functionality. No need to add to each class.